mongowire/wire_message.go

```diff
@@ -0,0 +1,100 @@
+package mongowire
+
+import (
+	"github.com/mongodb/ftdc/bsonx"
+	"github.com/pkg/errors"
+	"github.com/tychoish/mongorpc/model"
+)
+
+type opMessageSection interface {
+	Type() uint8
+	Name() string
+	DB() string
+	Documents() []*bsonx.Document
+	Serialize() []byte
+}
+
+type opMessagePayloadType0 struct {
+	PayloadType uint8
+	Document    *bsonx.Document
+}
+
+func (p *opMessagePayloadType0) Type() uint8                  { return 0 }
+func (p *opMessagePayloadType0) Name() string                 { return "" }
+func (p *opMessagePayloadType0) Documents() []*bsonx.Document { return []*bsonx.Document{p.Document} }
+func (p *opMessagePayloadType0) Serialize() []byte            { return nil }
+func (p *opMessagePayloadType0) DB() string {
+	key, err := p.Document.LookupErr("$db")
+	if err != nil {
+		return ""
+	}
+
+	val, ok := key.StringValueOK()
+	if !ok {
+		return ""
+	}
+
+	return val
+}
+
+type opMessagePayloadType1 struct {
+	PayloadType uint8
+	Size        int32
+	Identifer   string
+	Payload     []*bsonx.Document
+}
+
+func (p *opMessagePayloadType1) Type() uint8                  { return 1 }
+func (p *opMessagePayloadType1) Name() string                 { return p.Identifer }
+func (p *opMessagePayloadType1) DB() string                   { return "" }
+func (p *opMessagePayloadType1) Documents() []*bsonx.Document { return p.Payload }
+func (p *opMessagePayloadType1) Serialize() []byte            { return nil }
+
+func (m *opMessage) Header() MessageHeader { return m.header }
+func (m *opMessage) HasResponse() bool     { return m.Flags > 1 }
+func (m *opMessage) Scope() *OpScope {
+	return &OpScope{
+		Type: m.header.OpCode,
+	}
+}
+
+func (m *opMessage) Serialize() []byte { return nil }
+
+// TODO:
+//   - finish implementation of parseMsgMessageBody
+//   - implement message interface
+//      - Serialize
+
+func NewOpMessage(moreToCome bool, document *bsonx.Document, items ...model.SequenceItem) Message {
```

Comment:
This should be set on the `*opMessage`. Also, if there can be arbitrary number of sections, this be able to accept >1 document of kind 0.